add(explanation): Locks (#57)
canonical/real-time-ubuntu-docs: docs/explanation/locks.rst

The ideal type of spinlock to use depends on the specific contexts between which the data needs to be shared.

Within the class of spinlocks, there are various types:

.. _rwlock_t:

rwlock_t
~~~~~~~~

``rwlock_t`` (defined on `include/linux/rwlock_types.h`_)  is a variant of the read/write lock for spinning locks.
Consumers of this lock are divided into two classes: readers and writers.
This lock structure allows multiple readers but only one writer at a time.
By implementing this approach, the number of locks used can be reduced, which in turn decreases the overhead associated with locking.
This lock behaves different on ``PREEMPT_RT`` kernels.

- On non-PREEMPT_RT kernels it relies on the architecture dependent assembly implementation of a ``arch_rwlock_t``.
- On PREEMPT_RT kernels it is implemented as a ``rwbase_rt``, realizing a sleeping lock.

.. _spinlock_t:

spinlock_t
~~~~~~~~~~

``spinlock_t`` is the most generic implementation of spinlock structures, as defined in `include/linux/spinlock_types.h`_.
It is implemented in two different ways, depending on whether the real-time kernel configuration (i.e., with ``PREEMPT_RT`` enabled) is used.

> - In non-PREEMPT_RT scenarios, ``spinlock_t`` is a ``typedef`` structure based on :ref:`raw_spinlock_t`.

Reviewer: typedef by itself is not a structure, but an alias or identifier for another type, which could be a structure. 

In any case, best to avoid explaining code here:
```suggestion
- In non-PREEMPT_RT scenarios, ``spinlock_t`` maps to :ref:`raw_spinlock_t`.
```

- In PREEMPT_RT kernels, this is implemented as a :ref:`rt_mutex`, making this a sleeping lock.

.. _raw_spinlock_t:

raw_spinlock_t
~~~~~~~~~~~~~~

``raw_spinlock_t`` (defined on `include/linux/spinlock_types_raw.h`_) is a strict spinning lock implementation that operates consistently across all kernels, including real-time kernels.
It is reserved for critical core code, low-level interrupt handling, and scenarios where disabling preemption or interrupts is necessary.
For example, this lock is used to safely access hardware state.
In some cases, ``raw_spinlock_t`` can also be employed when the critical section is very small, thereby avoiding the overhead associated with :ref:`rt_mutex`.

mutex
~~~~~

A Mutex, short for "mutual exclusion" is used to protect critical sections of code, ensuring that only one thread or process can access a shared resource at any given time.

> Specifically, the ``mutex`` type (defined on `include/linux/mutex_types.h`_) is a simple mutex implementation that varies regarding if the current kernel is a PREEMPT_RT kernel or not.

Reviewer: ```suggestion
Specifically, the ``mutex`` type (defined on `include/linux/mutex_types.h`_) is a simple mutex implementation that varies depending on whether the current kernel is a PREEMPT_RT kernel or not.
```

> - For non-PREEMPT_RT kernels the implementation relies on an atomic operator which holds the owner and a :ref:`raw_spinlock_t`

Reviewer: ```suggestion
- For non-PREEMPT_RT kernels, the implementation relies on an atomic operator which holds the owner and a :ref:`raw_spinlock_t`
```

> - On PREEMPT_RT kernels, the implementation is based on a sleeping lock.

Reviewer: ```suggestion
- On PREEMPT_RT kernels, the implementation is based on :ref:`rt_mutex`.
```

The primary use case for this type of lock occurs when data needs to be shared between different user contexts.

.. _rt_mutex:

rt_mutex
~~~~~~~~~~

RT-mutexes (defined on `include/linux/rtmutex.h`_) are a special type of mutex that support priority inheritance.
These mutexes are particularly useful in real-time kernels because priority inheritance is a protocol designed to mitigate the effects of `priority inversion`_.
It's important to note that priority inheritance has limitations in non-real-time kernels due to preemption and interrupt-disabled sections.
Additionally, priority inheritance cannot take effect in regions of code where preemption or interrupts are disabled.
The benefits of priority inheritance are confined to regions where preemption is enabled (i.e., in preemptive task contexts).
Essentially, RT-mutexes can be used to implement spinlock_t and read-write locks, providing the added advantage of priority inheritance where applicable.

semaphore
~~~~~~~~~~

> The ``semaphore`` (defined in `include/linux/semaphore.h`_) structure discussed here follow the counting semaphore structure (as the binary semaphore equivalent is essentially a mutex).

Reviewer: ```suggestion
The ``semaphore`` (defined in `include/linux/semaphore.h`_) structure discussed here follows the counting semaphore structure (as the binary semaphore equivalent is essentially a mutex).
```

percpu_rw_semaphore
~~~~~~~~~~~~~~~~~~~

The ``percpu_rw_semaphore`` (defined in `include/linux/percpu-rwsem.h`_) is a modern read-write semaphore designed with optimization for read-heavy workloads.
Traditional read-write semaphores suffer from performance degradation when multiple cores acquire the lock for reading, as the cache line containing the semaphore bounces between the L1 caches of those cores.
With percpu_rw_semaphore, reading is extremely fast due to the use of RCU (`Read-Copy-Update`_), which eliminates the need for atomic instructions in the lock and unlock paths.
However, locking for writing is significantly more expensive, as it involves calling synchronize_rcu(), which can take hundreds of milliseconds.

CPU Local Locks
---------------

On non-PREEMPT_RT kernels, ``local_lock`` functions are wrappers around preemption and interrupt-disabling primitives.
Unlike other locking mechanisms, disabling preemption or interrupts is strictly a CPU-local concurrency control technique and is not suitable for managing inter-CPU concurrency.

local_lock_t
~~~~~~~~~~~~

> ``local_lock_t`` (defined in `include/linux/local_lock_internal.h`_) is a cpu lock lock implementation that is implemented in two different ways depending on whether the real-time kernel configuration (i.e., with ``PREEMPT_RT`` enabled) is used.

Reviewer: ```suggestion
``local_lock_t`` (defined in `include/linux/local_lock_internal.h`_) is a cpu lock implementation that is implemented in two different ways depending on whether the real-time kernel configuration (i.e., with ``PREEMPT_RT`` enabled) is used.
```

> - In non-preemptive scenarios it is just a simple typedef structure without any special fields.

Reviewer: Not sure this is accurate.

Here is the kernel doc:
> On non-PREEMPT_RT kernels, local_lock functions are wrappers around preemption and interrupt disabling primitives. Contrary to other locking mechanisms, disabling preemption or interrupts are pure CPU local concurrency control mechanisms and not suited for inter-CPU concurrency control.

Code includes two structs added conditionally:
https://github.com/torvalds/linux/blob/v6.8/include/linux/local_lock_internal.h#L11-L16

In any case, I'd avoid explaining code here (i.e. mentioning `typedef`).

> - With ``PREEMPT_RT`` enabled it maps to a :ref:`spinlock_t`

Reviewer: ```suggestion
- With ``PREEMPT_RT`` enabled, it maps to a per CPU :ref:`spinlock_t`, which protects the critical section while staying preemptible. 
```

Linux kernel locks
==================

.. note::

  This documentation covers some kernel internals specific to Linux kernel v6.8.
  The details provided here may not fully apply to newer kernel versions.
  
  The interface of kernel locks hasn't changed significantly over the years, nevertheless they are considered as unstable and subject to change.

>   This is discussed in the `"The Linux Kernel Driver Interface" document`_.

Reviewer: ```suggestion
  This is discussed in `The Linux Kernel Driver Interface`_  documentation.
```

- `Spinning locks`_
- `Sleeping locks`_ (also known as blocking locks)
- `CPU local locks`_

Spinlocks
----------

Spinlocks (found in `include/linux/spinlock.h`_) are locks that keep polling (or "spinning"), attempting to acquire the lock.
They are small, fast, and simple to use, making them generally applicable throughout the kernel.
Spinlocks are ideal for situations where the kernel process cannot sleep.
The ideal type of spinlock to use depends on the specific contexts between which the data needs to be shared.

Within the class of spinlocks, there are various types:

.. _rwlock_t:

rwlock_t
~~~~~~~~

``rwlock_t`` (defined on `include/linux/rwlock_types.h`_)  is a variant of the read/write lock for spinning locks.
Consumers of this lock are divided into two classes: readers and writers.
This lock structure allows multiple readers but only one writer at a time.
By implementing this approach, the number of locks used can be reduced, which in turn decreases the overhead associated with locking.
This lock behaves different on ``PREEMPT_RT`` kernels.

- On non-PREEMPT_RT kernels it relies on the architecture dependent assembly implementation of a ``arch_rwlock_t``.

> - On PREEMPT_RT kernels it is implemented to a ``rwbase_rt`` is implemented as a sleeping lock.

Reviewer: Need to reword, maybe:
```suggestion
- On PREEMPT_RT kernels it is implemented as a ``rwbase_rt``, realizing a sleeping lock.
```

.. _spinlock:

> spinlock_t

Reviewer: change ref to `spinlock_t`

With percpu_rw_semaphore, reading is extremely fast due to the use of RCU (`Read-Copy-Update`_), which eliminates the need for atomic instructions in the lock and unlock paths.
However, locking for writing is significantly more expensive, as it involves calling synchronize_rcu(), which can take hundreds of milliseconds.

CPU Local Locks
---------------

On non-PREEMPT_RT kernels, ``local_lock`` functions are wrappers around preemption and interrupt-disabling primitives.
Unlike other locking mechanisms, disabling preemption or interrupts is strictly a CPU-local concurrency control technique and is not suitable for managing inter-CPU concurrency.

local_lock_t
~~~~~~~~~~~~

``local_lock_t`` (defined in `include/linux/local_lock_internal.h`_) is a cpu lock lock implementation that is implemented in two different ways depending on whether the real-time kernel configuration (i.e., with ``PREEMPT_RT`` enabled) is used.

- In non-preemptive scenarios it is just a simple typedef structure without any special fields.
- With ``PREEMPT_RT`` enabled it maps to a :ref:`spinlock`

References
----------

> `Lock types and their rules`_
> 
> `Unreliable Guide To Locking - Rusty Russell`_

Reviewer: Change to bullets:
```suggestion
- `Lock types and their rules`_
- `Unreliable Guide To Locking - Rusty Russell`_
```

.. _spin_lock(): https://github.com/torvalds/linux/blob/v6.8/include/linux/spinlock.h#L349C29-L349C39
.. _spin_unlock(): https://github.com/torvalds/linux/blob/v6.8/include/linux/spinlock.h#L389
.. _spin_lock_bh(): https://github.com/torvalds/linux/blob/v6.8/include/linux/spinlock.h#L354
.. _spin_unlock_bh(): https://github.com/torvalds/linux/blob/v6.8/include/linux/spinlock.h#L394
.. _`spin_lock_irqsave()`: https://github.com/torvalds/linux/blob/v6.8/include/linux/spinlock.h#L379
.. _`spin_lock_irq()`: https://github.com/torvalds/linux/blob/v6.8/include/linux/spinlock.h#L374
.. _`spin_unlock_irq()`: https://github.com/torvalds/linux/blob/v6.8/include/linux/spinlock.h#L399
.. _`spin_unlock_irqrestore()`: https://github.com/torvalds/linux/blob/v6.8/include/linux/spinlock.h#L404
.. _`Unreliable Guide To Locking - Rusty Russell`: https://docs.kernel.org/kernel-hacking/locking.html
.. _`Lock types and their rules`: https://docs.kernel.org/locking/locktypes.html
.. _`system call`: https://en.wikipedia.org/wiki/System_call
.. _`completions`: https://docs.kernel.org/scheduler/completion.html
.. _`priority inversion`: https://en.wikipedia.org/wiki/Priority_inversion

> .. _`"The Linux Kernel Driver Interface" document`: https://docs.kernel.org/process/stable-api-nonsense.html

Reviewer: ```suggestion
.. _`The Linux Kernel Driver Interface`: https://docs.kernel.org/process/stable-api-nonsense.html
```

Linux kernel locks
==================

> .. note::

Author: Should this be a `warning` or either a `tip` instead?

Reviewer: I think a `note` is alright, referring to the specific kernel version used to write the documentation. 

However, I don't think we should be documenting kernel space interfaces. Are the documented locks and functions available from the user space and relevant to this RT Ubuntu's readers?

Author: It's an explanation, the user who end up reading this is willing to understand details of how it works, so from the point of view of what an **explanation** is supposed to be this is relevant.

Also, we already documented kernel space interfaces, for example, the schedulers one.

On `SMP systems`_ (Symmetric Multiprocessing), the Linux kernel provides various synchronization mechanisms to safeguard shared data structures from concurrent access, preventing `race conditions`_ in the kernel space.

Locks are primarily used to prevent race conditions, which are a concern in both SMP and real-time systems.
Even in a single-processor real-time system, the preemption mechanism can lead to race conditions similar to those found in an SMP system.

There are essentially three classes of locks in the kernel:

- `Spinning locks`_
- `Sleeping locks`_ (also known as blocking locks)
- `CPU local locks`_

Spinlocks
----------

Spinlocks (found in `include/linux/spinlock.h`_) are locks that keep polling (or "spinning"), attempting to acquire the lock.
They are small, fast, and simple to use, making them generally applicable throughout the kernel.
Spinlocks are ideal for situations where the kernel process cannot sleep.
The ideal type of spinlock to use depends on the specific contexts between which the data needs to be shared.

Within the class of spinlocks, there are various types:

.. _rw-lock:

> rwlock_t

Reviewer: Could we use the same names to avoid confusion?

Applies to a few other instances.

Author: Done! Thank you!

~~~~~~~~

``rwlock_t`` (defined on `include/linux/rwlock_types.h`_)  is a variant of the read/write lock for spinning locks.
Consumers of this lock are divided into two classes: readers and writers.
This lock structure allows multiple readers but only one writer at a time.

> By implementing this approach, the number of locks used can be reduced, which in turn decreases the overhead associated with locking.

Reviewer: I'd also document that this is converted to a sleeping lock on PREEMPT_RT kernels.

Author: Actually, it is not! It relies on an atomic and a [rwbase_rt](https://elixir.bootlin.com/linux/v6.10.6/C/ident/rwbase_rt) structure, which is implemented using an atomic and 
 a [rt_mutex_base](https://elixir.bootlin.com/linux/v6.10.6/C/ident/rt_mutex_base) and this last one is a `raw_spinlock_t`

Reviewer: I've stated that based on the kernel docs, [here](https://www.kernel.org/doc/html/latest/locking/locktypes.html):
> On PREEMPT_RT kernels, these lock types are converted to sleeping locks:
> - local_lock
> - spinlock_t
> - rwlock_t

Author: Oh, yeah, I think that it is true in the sense that it becomes a `rt_mutex_base` which is indeed a sleeping lock!

.. _spinlock:

spinlock_t
~~~~~~~~~~

``spinlock_t`` is the most generic implementation of spinlock structures, as defined in `include/linux/spinlock_types.h`_.
It is implemented in two different ways, depending on whether the real-time kernel configuration (i.e., with ``PREEMPT_RT`` enabled) is used.

> - In non-PREEMPT_RT scenarios, ``spinlock_t`` is a ``typedef`` structure based on :ref:`raw_spinlock`.

Reviewer: This is the default and under the equivalent section. The explanation is a bit redundant. 
```suggestion

```

Author: There are a bunch of other types that maps to a `raw_spinlock_t`, so I wouldn't say that this is redundant. 

Also, why to explain what happens in PREEMPT_RT scenarios but not on non-PREEMPT_RT? Wouldn't be better to leave, so the user can see the difference between those?

> - In preemptive scenarios (real-time kernels), ``spinlock_t`` is implemented as an :ref:`rt_mutex`, meaning it uses an ``rt_mutex`` in its ``typedef`` structure. This means that ``PREEMPT_RT`` changes the lock category of ``spinlock_t`` from a spinning lock to a sleeping lock.

Reviewer: ```suggestion
- In PREEMPT_RT kernels, this is implemented as a :ref:`rt_mutex`, making this a sleeping lock.
```

.. _raw_spinlock:

raw_spinlock_t
~~~~~~~~~~~~~~

``raw_spinlock_t`` (defined on `include/linux/spinlock_types_raw.h`_) is a strict spinning lock implementation that operates consistently across all kernels, including real-time kernels.
It is reserved for critical core code, low-level interrupt handling, and scenarios where disabling preemption or interrupts is necessary.
For example, this lock is used to safely access hardware state.
In some cases, ``raw_spinlock_t`` can also be employed when the critical section is very small, thereby avoiding the overhead associated with :ref:`rt_mutex`.

Spinlock operations

> ~~~~~~~~~~~~~~~~~~~

Reviewer: This heading is at the same level with spinlock types. Move all spinlock types to one level lower under its own subsection?

Reviewer: I'm wondering if these are in the scope of this document. Good explanations there, so maybe they should be moved to a separate document.

Author: I've changed Spinlock Operations to H2 and all the operations to H3.

Reviewer: Please remove from this version of the doc. They can be added together with the documentation of other operations.

Reviewer: Diff for removal: https://github.com/canonical/real-time-ubuntu-docs/compare/f205cd1d5499bf93b7da6c7e5b08a1c577cd6c34..6d2bd2e09a6011394d14c247daa226ca72b92727

Note that softIRQs (including tasklets and timers) run on return from hardware interrupts, so `spin_lock_irq()`_ also prevents these from executing.
In this sense, `spin_lock_irqsave()`_ is the most general and powerful locking function.
This locking mechanism is also used when data needs to be shared between two hardware IRQ handlers, a rare but possible scenario.
This lock is essentially a superset of all the previous spinlock primitives.
The counterpart to this lock is `spin_unlock_irqrestore()`_, which releases the lock and clean the flags variable.

Sleeping Locks (Blocking locks)
-------------------------------

Sleeping locks have the primary characteristic that tasks holding them can be put to sleep.
Unlike spinlocks, sleeping locks (also known as blocking locks) are used in kernel regions where the process context can sleep, typically in user contexts.

Mutexes
~~~~~~~

In the Linux kernel, mutexes are a specific type of locking primitive that enforces serialization on shared memory systems.
A mutex, short for "mutual exclusion," is used to protect critical sections of code, ensuring that only one thread or process can access a shared resource at any given time.

> The primary use case for this type of lock occurs when data needs to be shared between different user contexts.

Reviewer: The header and explanation make it sound like this section is explanining mutexes in general. But as far as I see from kernel docs `mutex` is a type of sleeping lock, alongside the rest below.

Author: Yeah, I've just added a better explanation for this one!

WW-Mutexes support flexible locking strategies, making them suitable for dynamic and unordered lock acquisition scenarios.

percpu_rw_semaphore
~~~~~~~~~~~~~~~~~~~

The ``percpu_rw_semaphore`` (defined in `include/linux/percpu-rwsem.h`_) is a modern read-write semaphore designed with optimization for read-heavy workloads.
Traditional read-write semaphores suffer from performance degradation when multiple cores acquire the lock for reading, as the cache line containing the semaphore bounces between the L1 caches of those cores.
With percpu_rw_semaphore, reading is extremely fast due to the use of RCU (`Read-Copy-Update`_), which eliminates the need for atomic instructions in the lock and unlock paths.
However, locking for writing is significantly more expensive, as it involves calling synchronize_rcu(), which can take hundreds of milliseconds.

CPU Local Locks
---------------

On non-PREEMPT_RT kernels, ``local_lock`` functions are wrappers around preemption and interrupt-disabling primitives.
Unlike other locking mechanisms, disabling preemption or interrupts is strictly a CPU-local concurrency control technique and is not suitable for managing inter-CPU concurrency.

> local_lock_t

Reviewer: Why this differs in the kernel docs? There it is referred to as `local_lock`.

Author: The first mentions the `functions` the second mentions the lock type `local_lock_t`.

This allows the same code to be used inside a hardware IRQ handler (where interrupts are already off) and in softIRQs (where interrupt disabling is required).
Note that softIRQs (including tasklets and timers) run on return from hardware interrupts, so `spin_lock_irq()`_ also prevents these from executing.
In this sense, `spin_lock_irqsave()`_ is the most general and powerful locking function.
This locking mechanism is also used when data needs to be shared between two hardware IRQ handlers, a rare but possible scenario.
This lock is essentially a superset of all the previous spinlock primitives.
The counterpart to this lock is `spin_unlock_irqrestore()`_, which releases the lock and clean the flags variable.

Sleeping Locks (Blocking locks)
-------------------------------

Sleeping locks, also known as blocking locks, have the primary characteristic that tasks holding them can be put to sleep.
Unlike spinlocks, sleeping locks are used in kernel regions where the process context can sleep, typically in user contexts.
Most sleeping locks in the Linux kernel are mutexes, a specific type of locking primitive designed to enforce serialization on shared memory systems.

mutex
~~~~~

> A mutex, short for "mutual exclusion," is used to protect critical sections of code, ensuring that only one thread or process can access a shared resource at any given time.

Reviewer: ```suggestion
A Mutex, short for "mutual exclusion" is used to protect critical sections of code, ensuring that only one thread or process can access a shared resource at any given time.
```

Specificaly, the ``mutex`` type (defined on `include/linux/mutex_types.h`_) is a simple mutex implementation that varies regarging if the current kernel is a PREEMPT_RT kernel or not.

> - For non-PREEMPT_RT kernels scenarios the implementation relies on an atomic which holds the owner and a :ref:`raw_spinlock_t`

Reviewer: ```suggestion
- For non-PREEMPT_RT kernels the implementation relies on an atomic which holds the owner and a :ref:`raw_spinlock_t`
```

Instead of just "atomic", say "atomic operation"?

Reviewer: Instead of just "atomic", say "atomic operation"?

Author: I would say that it is an "atomic operator", because, I'm referring to this:

https://github.com/torvalds/linux/blob/e8f897f4afef0031fe618a8e94127a0934896aba/include/linux/mutex_types.h#L41-L43
